### PR TITLE
subsys: settings: Implemented simple RAM settings backend

### PR DESCRIPTION
**Pull request by ArekBalysNordic**
Currently, the Zephyr settings do not support the RAM backend.
Adding the RAM backend allows storing settings in volatile memory that can be useful for testing purposes. For example, in CI testing, a large amount of data is stored and erased from the NVS memory many times. This solution can increase the Flash lifetime in testing.

The simple RAM settings implementation allows using the ram (and heap) as a settings backend. All settings saved using Zephyr's API
are saved into slist, and each value of that list is created dynamically during entry creation according to a specific length.

RAM backend can be activated by setting CONFIG_SETTINGS_RAM=y Zephyr's kconfig.

Added unit tests for RAM settings backend in Zephyr tests/subsys.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

### CODE REVIEW

File: subsys/settings/Kconfig [hunk 1]
```diff
@@ -35,6 +35,7 @@ config SETTINGS_ENCODE_LEN
 
 choice SETTINGS_BACKEND
 	prompt "Storage back-end"
+	default SETTINGS_RAM
```
Comment (pdgendt) [inline subsys/settings/Kconfig:38]: The current default `SETTINGS_NONE` should either be removed or kept as the default. IMO `SETTINGS_NONE` should be kept, making the RAM backend an explicit choice for the user.

Also, as the RAM backend would be mostly used for testing purposes, it makes sense to print out a warning during the build.

File: subsys/settings/src/settings_ram.c [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+		if (entry) {
+			if (memcmp(entry->key, name, entry->key_len) == 0) {
+				if (entry->value) {
+					/* free memory for the entry value */
+					k_free(entry->value);
+				}
+				/* free memory for the entry */
+				k_free(entry);
+				LOG_INF("Deleting entry: %s with size: %zu", name, entry->key_len + entry->value_len);
+				return sys_slist_find_and_remove(&settings_list, &entry->node);
```
Comment (pdgendt) [inline subsys/settings/src/settings_ram.c:144]: `entry` is freed and used afterwards.

File: subsys/settings/src/settings_ram.c [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+	new_entry = k_malloc(sizeof(struct settings_ram_entry));
+	if (!new_entry) {
+		LOG_ERR("Could not allocate memory for new entry");
+		return false;
+	}
+
+	new_entry->value = k_malloc(len);
+	if (!new_entry->value) {
+		LOG_ERR("Could not allocate memory for new entry value");
+		return false;
```
Comment (pdgendt) [inline subsys/settings/src/settings_ram.c:193]: You need to free `new_entry` before returning here.

**Review by pdgendt — CHANGES_REQUESTED** [review]

**Comment by pdgendt** [comment]
Please amend any changes made from reviews to the existing commits, instead of creating a new commit for it. Simply squash (or do a fancy interactive rebase if more than 1 commit is inside the PR) and force push.

Also no need to add usernames to commit messages :)

**Comment by ArekBalysNordic** [comment]
@pdgendt Ok, done :)

File: subsys/settings/src/settings_ram.c [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+ * @return true if entry has been deleted
+ * @return false entry has not been found
+ */
+static bool delete_entry(const char *name)
+{
+	struct settings_ram_entry *entry = NULL;
+	bool rc = false;
+	SYS_SLIST_FOR_EACH_CONTAINER (&settings_list, entry, node) {
+		if (entry) {
+			if (memcmp(entry->key, name, entry->key_len) == 0) {
```
Comment (pdgendt) [inline subsys/settings/src/settings_ram.c:137]: Can the `entry` variable ever be NULL? Either way, these conditions could be checked in a single statement.

```suggestion
		if (entry != NULL && memcmp(entry->key, name, entry->key_len) == 0) {
```

File: subsys/settings/src/settings_ram.c [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+				if (entry->value) {
+					/* free memory for the entry value */
+					k_free(entry->value);
+				}
+				/* free memory for the entry */
+				rc = sys_slist_find_and_remove(&settings_list, &entry->node);
+				k_free(entry);
+				LOG_INF("Deleting entry: %s with size: %zu", name, entry->key_len + entry->value_len);
```
Comment (pdgendt) [inline subsys/settings/src/settings_ram.c:145]: `entry` is still being used after being freed in the `LOG_INF` statement.

**Review by pdgendt — CHANGES_REQUESTED** [review]
Still a `k_free` issue to be fixed.

General question, is it preferred to use the system zephyr heap, or should a specific one be added, with a configurable size from `Kconfig`?

**Comment by ArekBalysNordic** [comment]
I thought about it this way the RAM settings backend uses Zephyr's heap because settings belong to Zephyr. Still, after your consideration, I think it will be nice to add a specific one, and then SETTINGS_RAM Kconfig will require providing SETTINGS_RAM_HEAP_SIZE Kconfig.

**Comment by Laczen** [comment]
@ArekBalysNordic, what is the advantage over using the flash simulator driver in RAM?

**Comment by ArekBalysNordic** [comment]
@Laczen The main advantage over storing Zephyr's settings in RAM is reducing Flash's wear during testing. Currently, I work on the Matter stack, and we have a lot of nightly tests that perform many erasing and writing the flash memory due to factory resets. In addition, we test mainly a network stack, so we don't think a lot about a settings backend, so we started considering moving settings from Flash to RAM only for testing purposes.

File: subsys/settings/src/settings_ram.c [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+ * @return false if could not find given entry
+ */
+static bool get_entry(const char *name, char *value, size_t *len)
+{
+	if (!name || !value || !len) {
+		return false;
+	}
+	struct settings_ram_entry *entry = NULL;
+	SYS_SLIST_FOR_EACH_CONTAINER (&settings_list, entry, node) {
+		if (strncmp(entry->key, name, strnlen(entry->key, SETTINGS_MAX_NAME_LEN)) == 0) {
```
Comment (pdgendt) [inline subsys/settings/src/settings_ram.c:98]: No need to use `strnlen` here (same for `try_to_update_entry` and `delete_entry`

```suggestion
		if (strncmp(entry->key, name, SETTINGS_MAX_NAME_LEN) == 0) {
```

File: subsys/settings/src/settings_ram.c [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+	new_entry->value = k_malloc(len);
+	if (!new_entry->value) {
+		LOG_ERR("Could not allocate memory for new entry value");
+		k_free(new_entry);
+		return false;
+	}
+
+	memset(new_entry->key, 0, SETTINGS_MAX_NAME_LEN);
+	memcpy(new_entry->key, name, strnlen(name, SETTINGS_MAX_NAME_LEN) + 1);
```
Comment (pdgendt) [inline subsys/settings/src/settings_ram.c:187]:
```suggestion
	strncpy(new_entry->key, name, SETTINGS_MAX_NAME_LEN);
```

**Review by pdgendt — CHANGES_REQUESTED** [review]

**Comment by Laczen** [comment]
> @Laczen The main advantage over storing Zephyr's settings in RAM is reducing Flash's wear during testing. Currently, I work on the Matter stack, and we have a lot of nightly tests that perform many erasing and writing the flash memory due to factory resets. In addition, we test mainly a network stack, so we don't think a lot about a settings backend, so we started considering moving settings from Flash to RAM only for testing purposes. I guess it can also be helpful for other teams that work on the high abstraction level on Embedded devices.

I agree with the need to reduce flash wear during testing. The flash simulator is a simulation of flash in RAM. Just configure and use instead of making a new backend.

**Comment by ArekBalysNordic** [comment]
@Laczen Where can I find documentation of the mentioned flash simulator? I wanted to try it, but I can see that setting the CONFIG_FLASH_SIMULATOR Kconfig is insufficient.

**Comment by ArekBalysNordic** [comment]
For now, I've added a specific heap for RAM settings backend and Kconfig SETTINGS_RAM_HEAP_SIZE to configure it. I'm going to check how the flash simulator works, and if it meets my expectations, I agree this PR isn't necessary.

File: subsys/settings/Kconfig [hunk 8]
```diff
@@ -28,6 +28,15 @@ config SETTINGS_DYNAMIC_HANDLERS
 	help
 	  Enables the use of dynamic settings handlers
 
+config SETTINGS_RAM_HEAP_SIZE
+	int "heap size dedicated to RAM backend"
+	default 1024
+	depends on SETTINGS
```
Comment (pdgendt) [inline subsys/settings/Kconfig:34]:
```suggestion
	depends on SETTINGS_RAM
```

File: subsys/settings/src/settings_ram.c [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+#include <sys/sys_heap.h>
+#include <stdalign.h>
+
+LOG_MODULE_DECLARE(settings, CONFIG_SETTINGS_LOG_LEVEL);
+
+/* Alignment of memory blocks returned by malloc. */
+const static size_t kMallocAlignment = alignof(long long);
+uint8_t sHeapMemory[CONFIG_SETTINGS_RAM_HEAP_SIZE];
+struct sys_heap sHeap;
+SYS_MUTEX_DEFINE(heap_lock);
```
Comment (pdgendt) [inline subsys/settings/src/settings_ram.c:24]: Use the zephyr heap abstractions, no need for a mutex then.

```suggestion
static K_HEAP_DEFINE(ram_settings_heap, CONFIG_SETTINGS_RAM_HEAP_SIZE);
```

**Review by pdgendt — CHANGES_REQUESTED** [review]

**Comment by Laczen** [comment]
@ArekBalysNordic, to use the flash simulator for storage you should define an overlay file like (you can change the sizes according to your needs):

```
/delete-node/ &storage_partition;

/ {
	soc {
		flash_controller2: flash-controller@2 {
			compatible = "zephyr,sim-flash";
			reg = <0x00000000 DT_SIZE_K(40)>;

			#address-cells = <1>;
			#size-cells = <1>;
			erase-value = <0xff>;

			label = "flash_ctrl";

			flash_sim0: flash_sim@0 {
				status = "okay";
				compatible = "soc-nv-flash";
				label = "simulated_flash";
				erase-block-size = <4096>;
				write-block-size = <1>;
				reg = <0x00000000 DT_SIZE_K(40)>;

				partitions {
					compatible = "fixed-partitions";
					#address-cells = <1>;
					#size-cells = <1>;

					storage_partition: partition@0 {
						label = "storage";
						reg = <0x00000000 0x00000A000>;
					};
				};
			};
		};
	};
};
```

**Comment by Laczen** [comment]
> Is it even possible to run it on a physical device such as NRF52 in this way? I think the overlay you mentioned is proper only for Posix systems, but I would like to use it on NRF52, for example. I think it isn't trivial to add it to the NRF52 and simulate flash there. So my next advantage of using RAM settings backend is to simplify saving Zephyr settings on NRF52 without overlaying dts files. My intention with creating this PR was to store Zephyr's settings to RAM instead of Flash for Arm embedded devices, not POSIX systems. Please, correct me if I am wrong, and I can use the flash simulator in my projects. Still, I can't find documentation describing how to use the flash simulator. How can I get to know how you generated this overlay?

No it is not for posix systems only, the flash simulator requires a new device that is setup like in the above overlay file. The example overlay file is mostly coming from mcuboot where this setup is used to place an extra image in ram. Using a overlay file (combined with the option to enable flash_simulator device) during testing to reduce flash wear seems to me a very good way not to change a lot of things during the test. It is setup to just redirect the flash writing to ram, nothing else is changed.

**Review by nvlsianpu — COMMENTED** [review]
I think @Laczen's input on possibility of using flash simulator is worth to be considered as an alternative to this.
Of course usage of the RAM backend is simpler to be configured and it does not involve FS/storage and flash driver - but it looks like this is not an objective for this PR.
Something as CONFIG_SETTINGS_RAM_DEBUG might be introduced to a backend we have already.

New backend gives more code to be maintained - please consider my remarks.

**Comment by ArekBalysNordic** [comment]
I have checked the Flash simulator, and it works for me. However, I am a bit worried because this solution requires changing an overlay file for each device separately, whereas my idea is based on a single Kconfig. Moreover, I think there is a lack of documentation for using that Flash Simulator because I didn't find anything, and also anyone didn't give any. Basing on the example stored somewhere in mcuboot is very confusing. One way or another, I'm going to close this PR and use the Flash simulator in my projects.

**Comment by nvlsianpu** [comment]
@ArekBalysNordic I think you can create one DTC overlay and add it to various build using for either commandline, tests.yaml or CMakeLists.txt.

**Comment by Laczen** [comment]
@ArekBalysNordic, the lack of documentation is mainly because this is something rather unique, it is a device that is being setup and devices documentation in general is rather sparse. Devices are mostly defined by the dts for a specific soc and are inherent to this soc. The flash simulator is not inherent to a specific soc but more general. It would be possible to add some documentation in the flash section but nobody has volunteered yet to add this information. The usage is also quite limited mainly to testing and special cases (like the mcuboot case) as almost nobody has spare RAM to use the flash simulator in general.

To give this some perspective: at the time the flash simulator was added there was no (or limited) native_posix support so testing and developing a flash filesystem was very hard on flash wear, the simulator allowed to avoid this flash wear. Also at that time documentation was sparse.

Creating an overlay file to add the device to the soc (which is what is done in the example I gave) is a quite general way to extend whatever soc with specific devices that can be added optionally to a board.

If you enjoy the use of the flash simulator and think its use can be promoted with additional documentation please do not hesitate to open a PR for extra documentation.

**Pull request closed without merging**